test/threads.jl

```diff
@@ -141,3 +141,41 @@ end
 @test_throws TypeError Atomic{Bool}
 @test_throws TypeError Atomic{BigInt}
 @test_throws TypeError Atomic{Float64}
+
+# Test atomic fences
+type CommBuf
+    barrier::Atomic{Int}
+    data::Vector{Float64}
+    correct::Bool
+    CommBuf() = new(Atomic{Int}(0), Vector{Float64}(100), false)
+end
+function test_fence_write(commbuf::CommBuf)
+    sleep(0.1)
+    for i in eachindex(commbuf.data)
+        commbuf.data[i] = i
+    end
+    atomic_fence()
+    commbuf.barrier[] = 1
```

Comment:
Atomic fences on different threads are only ordered with respect to each other if there are ordered atomic writes/reads in between. The fence then orders the preceding non-atomic writes (on one thread) and the following non-atomic reads (on the other thread).

Unfortunately, on x86, the fence is a no-op, since the hardware always makes this guarantee anyway.